File: Sandu Mihai-Madalin/RadixSort.py
import math
import random
import numpy as np

def radixSort(arr):
    length = len(arr)
    maxN = max(arr)
    exp = 1
    while maxN // exp > 0:
        output = [0] * length
        count = [0] * 10

        for i in range(length):
            index = arr[i] // exp
            count[index % 10] += 1

        for i in range(0, 9):
            count[i + 1] += count[i]

        for i in range(length - 1, -1, -1):
            index = arr[i] // exp
            output[count[index % 10] - 1] = arr[i]
            count[index % 10] -= 1

        for i in range(length):
            arr[i] = output[i]
        exp *= 10


def main():
    arr = []
    inp1 = input("Introduce un nr (nr ales * 2 v-a fi valoarea maxima posibila in lista): ")
    for i in range(int(inp1)):
        arr.append(random.randint(1, int(inp1)*2))
    radixSort(arr)
    print(arr)

main()

File: proiectarea algoritmilor/Balan Bianca-Ioana/radix sort.cpp
#include <iostream>
#include <vector>
using namespace std;

void countingSort(int arr[], int n, int exp) {
    int output[n];
    int count[10] = {0};
    for (int i = 0; i < n; i++)
        count[(arr[i] / exp) % 10]++;

    for (int i = 1; i < 10; i++)
        count[i] += count[i - 1];

    for (int i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
    }

    for (int i = 0; i < n; i++)
        arr[i] = output[i];
}
void radixSort(int arr[], int n) {
    int maxElement = arr[0];
    for (int i = 1; i < n; i++)
        if (arr[i] > maxElement)
            maxElement = arr[i];

    for (int exp = 1; maxElement / exp > 0; exp *= 10)
        countingSort(arr, n, exp);
}

int main() {
    int arr[] = {170, 45, 75, 90, 802, 24, 2, 66};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Vectorul initial: ";
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;

    radixSort(arr, n);

    cout << "Vectorul sortat: ";
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;

    return 0;
}
